src/methods/destructuring.js

```diff
@@ -0,0 +1,434 @@
+import type { LexicalEnvironment } from "../environment.js";
+import { Reference } from "../environment.js";
+import type { PropertyKeyValue } from "../types.js";
+import { Value, ObjectValue, UndefinedValue } from "../values/index.js";
+import { AbruptCompletion } from "../completions.js";
+import { EvalPropertyName } from "../evaluators/ObjectExpression.js";
+import {
+  RequireObjectCoercible,
+  GetIterator,
+  IteratorClose,
+  IteratorStep,
+  IteratorValue,
+  GetValue,
+  IsAnonymousFunctionDefinition,
+  IsIdentifierRef,
+  HasOwnProperty,
+  SetFunctionName,
+  GetReferencedName,
+  PutValue,
+  ArrayCreate,
+  CreateDataProperty,
+  GetV,
+} from "./index.js";
+import type {
+  BabelNodeLVal,
+  BabelNodeArrayPattern,
+  BabelNodeObjectPattern,
+} from "babel-types";
+
+// ECMA262 12.15.5.2
+export function DestructuringAssignmentEvaluation(realm: Realm, pattern: BabelNodeArrayPattern | BabelNodeObjectPattern, value: Value, strictCode: boolean, env: LexicalEnvironment) {
+  if (pattern.type === "ObjectPattern") {
+    // 1. Perform ? RequireObjectCoercible(value).
+    RequireObjectCoercible(realm, value);
+
+    // 2. Return the result of performing DestructuringAssignmentEvaluation for AssignmentPropertyList using value as the argument.
+
+    for (let property of pattern.properties) {
+      // 1. Let name be the result of evaluating PropertyName.
+      let name = EvalPropertyName(property, env, realm, strictCode);
+
+      // 2. ReturnIfAbrupt(name).
+
+      // 3. Return the result of performing KeyedDestructuringAssignmentEvaluation of AssignmentElement with value and name as the arguments.
+      KeyedDestructuringAssignmentEvaluation(realm, property.value, value, name, strictCode, env);
+    }
+  } else if (pattern.type === "ArrayPattern") {
+    // 1. Let iterator be ? GetIterator(value).
+    let iterator = GetIterator(realm, value);
+
+    // 2. Let iteratorRecord be Record {[[Iterator]]: iterator, [[Done]]: false}.
+    let iteratorRecord = {
+      $Iterator: iterator,
+      $Done: false,
+    };
+
+    // 3. Let result be the result of performing IteratorDestructuringAssignmentEvaluation of AssignmentElementList using iteratorRecord as the argument.
+    let result;
+    try {
+      result = IteratorDestructuringAssignmentEvaluation(realm, pattern.elements, iteratorRecord, strictCode, env);
+    } catch (error) {
+      // 4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator, result).
+      if (iteratorRecord.$Done === false && error instanceof AbruptCompletion) {
+        throw IteratorClose(realm, iterator, error);
+      }
+      throw error;
+    }
+
+    // 4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator, result).
+    // TODO: Why does this work?
+    if (iteratorRecord.$Done === false) {
+      let completion = new AbruptCompletion(new UndefinedValue(realm));
+      let actualCompletion = IteratorClose(realm, iterator, completion);
```

Comment:
IteratorClose should be generalized to work on Completion, including NormalCompletion. When it was written originally, we didn't have a NormalCompletion type yet. Then you can wrap the result into a NormalCompletion, and throw only when IteratorClose returns a non-NormalCompletion, and otherwise return what's in it.